pkg/webhook/workloadspread/validating/workloadspread_validation.go

```diff
@@ -0,0 +1,253 @@
+				}
+				subSetNames.Insert(subsetName)
+			}
+
+			// need requiredNodeSelectorTerm or preferredNodeSelectorTerms except the end subset.
+			if subset.RequiredNodeSelectorTerm == nil && subset.PreferredNodeSelectorTerms == nil && i != len(subsets)-1 {
+				allErrs = append(allErrs, field.Invalid(fldPath.Index(i).Child("requiredNodeSelectorTerm"), subset.RequiredNodeSelectorTerm, "empty RequiredNodeSelectorTerm and PreferredNodeSelectorTerms is not valid for WorkloadSpread"))
+			} else {
+				if subset.RequiredNodeSelectorTerm != nil {
+					coreNodeSelectorTerm := &core.NodeSelectorTerm{}
+					if err := corev1.Convert_v1_NodeSelectorTerm_To_core_NodeSelectorTerm(subset.RequiredNodeSelectorTerm.DeepCopy(), coreNodeSelectorTerm, nil); err != nil {
+						allErrs = append(allErrs, field.Invalid(fldPath.Index(i).Child("requiredNodeSelectorTerm"), subset.RequiredNodeSelectorTerm, fmt.Sprintf("Convert_v1_NodeSelectorTerm_To_core_NodeSelectorTerm failed: %v", err)))
+					} else {
+						allErrs = append(allErrs, corevalidation.ValidateNodeSelectorTerm(*coreNodeSelectorTerm, fldPath.Index(i).Child("requiredNodeSelectorTerm"))...)
+					}
+				}
+
+				if subset.PreferredNodeSelectorTerms != nil {
+					corePreferredSchedulingTerms := make([]core.PreferredSchedulingTerm, 0, len(subset.PreferredNodeSelectorTerms))
+					for i, term := range subset.PreferredNodeSelectorTerms {
+						corePreferredSchedulingTerm := &core.PreferredSchedulingTerm{}
+						if err := corev1.Convert_v1_PreferredSchedulingTerm_To_core_PreferredSchedulingTerm(term.DeepCopy(), corePreferredSchedulingTerm, nil); err != nil {
+							allErrs = append(allErrs, field.Invalid(fldPath.Index(i).Child("preferredSchedulingTerms"), subset.PreferredNodeSelectorTerms, fmt.Sprintf("Convert_v1_PreferredSchedulingTerm_To_core_PreferredSchedulingTerm failed: %v", err)))
+						} else {
+							corePreferredSchedulingTerms = append(corePreferredSchedulingTerms, *corePreferredSchedulingTerm)
+						}
+					}
+
+					allErrs = append(allErrs, corevalidation.ValidatePreferredSchedulingTerms(corePreferredSchedulingTerms, fldPath.Index(i).Child("preferredSchedulingTerms"))...)
+				}
+			}
+
+			if subset.Tolerations != nil {
+				var coreTolerations []core.Toleration
+				for i, toleration := range subset.Tolerations {
+					coreToleration := &core.Toleration{}
+					if err := corev1.Convert_v1_Toleration_To_core_Toleration(&toleration, coreToleration, nil); err != nil {
+						allErrs = append(allErrs, field.Invalid(fldPath.Index(i).Child("tolerations"), subset.Tolerations, fmt.Sprintf("Convert_v1_Toleration_To_core_Toleration failed: %v", err)))
+					} else {
+						coreTolerations = append(coreTolerations, *coreToleration)
+					}
+				}
+				allErrs = append(allErrs, corevalidation.ValidateTolerations(coreTolerations, fldPath.Index(i).Child("tolerations"))...)
+			}
+
+			if subset.MaxReplicas != nil {
```

Comment:
is it valid if ScheduleStrategy is fixed and max replicas is not specified

Reply:
It should be valid.

Reply:
Because the maxReplicas is nil means no replica limits, we only permit the end subset' maxReplicas is nil.
If the nil subset is not the last, the back subset is invalid.

Reply:
1. all subset maxReplicas must be the same type: int or percent
2. only last subset max replicas can be not specified
2. Fixed
- if all maxreplicas is specified as percent, the total maxreplicas must equals 1, except the last subset is not specified.
3. Adaptive
- the last subset must be not specified